### ISSUE TITLE
TensorFlow `tf.function(jit_compile=False)` and eager execution start from nearly identical first-step metrics but diverge over some ResNet50 Adam steps

### ISSUE THREAD

**Issue by whitephilomel** [tensorflow/tensorflow#120996]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.18.0

### Custom code

Yes

### OS platform and distribution

Linux 5.15.0-113-generic x86_64, glibc 2.35

### Python version

3.11.15

### GCC/compiler version

GCC 14.3.0

### CUDA/cuDNN version

CUDA 12.5.1 / cuDNN 9

### GPU model and memory

Tesla V100S-PCIE-32GB

### Current behavior?

The same fixed ResNet50 training workload is run twice:
- graph arm: the train step is wrapped by `tf.function(jit_compile=False)`;
- eager arm: the same Python train step runs eagerly.
Both arms start from the same copied initial weights and consume the same deterministic synthetic batch sequence. The first step is nearly identical, but the two trajectories diverge over 20 optimizer steps.
Observed final result:

| arm | final loss | final grad norm |
|---|---:|---:|
| graph, `tf.function(jit_compile=False)` | 5.515571594238281 | 8.42819881439209 |
| eager | 3.878284454345703 | 4.26118278503418 |

The important detail is that this does not start as a completely different computation. Step 1 is almost identical:

| arm | step 1 loss | step 20 loss | step 1 grad norm | step 20 grad norm |
|---|---:|---:|---:|---:|
| graph | 4.083001136779785 | 5.515571594238281 | 853.9214477539062 | 8.42819881439209 |
| eager | 4.082990646362305 | 3.878284454345703 | 853.6309204101562 | 4.26118278503418 |

## Expected behavior

For the same model, initial weights, deterministic inputs, labels, optimizer, and `jit_compile=False`, eager execution and `tf.function` graph execution should remain numerically close. Small GPU floating-point differences are expected, but a 29.7% final-loss difference and 49.4% final-gradient-norm difference look too large for this graph/eager execution-mode change alone.

### Standalone code to reproduce the issue

```shell
Run:


python tensorflow_repro.py \
  --profile confirmed \
  --steps 20 \
  --checks graph_vs_eager \
  --cuda-visible-devices 0

tensorflow_repro.py

#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import math
import os
import random
import time
from pathlib import Path
from typing import Any


LOSS_REL_THRESHOLD = 0.05
GRAD_REL_THRESHOLD = 0.10


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Standalone TensorFlow graph/eager and XLA/non-XLA differential reproducer."
    )
    parser.add_argument(
        "--profile",
        choices=["confirmed", "original_config"],
        default="confirmed",
        help=(
            "confirmed: matches the reduced confirmed replay; "
            "original_config: closer to the original boundary config."
        ),
    )
    parser.add_argument("--checks", default="graph_vs_eager,xla_vs_nonxla")
    parser.add_argument("--steps", type=int, default=None)
    parser.add_argument("--seed", type=int, default=None)
    parser.add_argument("--batch-size", type=int, default=None)
    parser.add_argument("--grad-accum-steps", type=int, default=None)
    parser.add_argument("--learning-rate", type=float, default=None)
    parser.add_argument("--weight-decay", type=float, default=None)
    parser.add_argument("--precision-policy", default=None)
    parser.add_argument("--image-size", type=int, default=224)
    parser.add_argument("--num-classes", type=int, default=10)
    parser.add_argument("--cuda-visible-devices", default="0")
    parser.add_argument("--intra-op-threads", type=int, default=2)
    parser.add_argument("--inter-op-threads", type=int, default=8)
    parser.add_argument("--loss-rtol", type=float, default=LOSS_REL_THRESHOLD)
    parser.add_argument("--grad-rtol", type=float, default=GRAD_REL_THRESHOLD)
    parser.add_argument("--output-dir", default=None)
    parser.add_argument("--fail-on-divergence", action="store_true")
    parser.add_argument("--print-step-values", action="store_true")
    return parser.parse_args()


def profile_defaults(profile: str) -> dict[str, Any]:
    if profile == "original_config":
        return {
            "steps": 20,
            "seed": 1,
            "batch_size": 8,
            "grad_accum_steps": 16,
            "learning_rate": 0.05,
            "weight_decay": 0.01,
            "precision_policy": "mixed_bfloat16",
        }
    return {
        "steps": 20,
        "seed": 1,
        "batch_size": 8,
        "grad_accum_steps": 1,
        "learning_rate": 0.05,
        "weight_decay": 0.0,
        "precision_policy": "float32",
    }


def apply_defaults(args: argparse.Namespace) -> argparse.Namespace:
    defaults = profile_defaults(args.profile)
    for key, value in defaults.items():
        attr = key.replace("-", "_")
        if getattr(args, attr) is None:
            setattr(args, attr, value)
    args.steps = int(args.steps)
    args.seed = int(args.seed)
    args.batch_size = int(args.batch_size)
    args.grad_accum_steps = int(args.grad_accum_steps)
    args.learning_rate = float(args.learning_rate)
    args.weight_decay = float(args.weight_decay)
    args.precision_policy = str(args.precision_policy)
    return args


def setup_output_dir(args: argparse.Namespace) -> Path:
    if args.output_dir:
        out = Path(args.output_dir).expanduser().resolve()
    else:
        stamp = time.strftime("%Y%m%d%H%M%S")
        out = Path.cwd() / f"standalone_tf_differential_{args.profile}_{args.steps}step_run_{stamp}"
    out.mkdir(parents=True, exist_ok=True)
    return out


def configure_runtime(args: argparse.Namespace) -> Any:
    os.environ["CUDA_VISIBLE_DEVICES"] = str(args.cuda_visible_devices)
    os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "1")
    os.environ.setdefault("TF_FORCE_GPU_ALLOW_GROWTH", "true")

    import numpy as np
    import tensorflow as tf

    random.seed(args.seed)
    np.random.seed(args.seed)
    tf.keras.utils.set_random_seed(args.seed)

    try:
        tf.config.threading.set_intra_op_parallelism_threads(int(args.intra_op_threads))
        tf.config.threading.set_inter_op_parallelism_threads(int(args.inter_op_threads))
    except RuntimeError:
        pass

    for gpu in tf.config.list_physical_devices("GPU"):
        try:
            tf.config.experimental.set_memory_growth(gpu, True)
        except RuntimeError:
            pass

    tf.keras.mixed_precision.set_global_policy(args.precision_policy)
    return tf


def make_model(tf: Any, args: argparse.Namespace) -> Any:
    tf.keras.utils.set_random_seed(args.seed)
    base = tf.keras.applications.ResNet50(
        weights=None,
        include_top=False,
        input_shape=(int(args.image_size), int(args.image_size), 3),
        pooling="avg",
    )
    outputs = tf.keras.layers.Dense(int(args.num_classes))(base.output)
    return tf.keras.Model(base.input, outputs)


def make_optimizer(tf: Any, args: argparse.Namespace) -> Any:
    try:
        return tf.keras.optimizers.Adam(
            learning_rate=float(args.learning_rate),
            weight_decay=float(args.weight_decay),
        )
    except TypeError:
        return tf.keras.optimizers.Adam(learning_rate=float(args.learning_rate))


def fixed_batch(tf: Any, args: argparse.Namespace, step: int) -> tuple[Any, Any]:
    seed = int(args.seed) + int(step)
    generator = tf.random.Generator.from_seed(seed)
    x = generator.normal(
        (
            int(args.grad_accum_steps),
            int(args.batch_size),
            int(args.image_size),
            int(args.image_size),
            3,
        ),
        dtype=tf.float32,
    )
    y_one = tf.range(int(args.batch_size), dtype=tf.int32) % int(args.num_classes)
    y = tf.tile(tf.expand_dims(y_one, axis=0), [int(args.grad_accum_steps), 1])
    return x, y


def clone_initial_weights(weights: list[Any]) -> list[Any]:
    return [w.copy() for w in weights]


def train_arm(
    tf: Any,
    args: argparse.Namespace,
    *,
    name: str,
    use_function: bool,
    jit_compile: bool,
    initial_weights: list[Any],
) -> dict[str, Any]:
    tf.keras.backend.clear_session()
    tf.keras.mixed_precision.set_global_policy(args.precision_policy)
    tf.keras.utils.set_random_seed(args.seed)

    model = make_model(tf, args)
    model.set_weights(clone_initial_weights(initial_weights))
    optimizer = make_optimizer(tf, args)
    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

    trainable = model.trainable_variables

    def step_fn(xs: Any, ys: Any) -> tuple[Any, Any]:
        accum_grads = [tf.zeros_like(v) for v in trainable]
        total_loss = tf.constant(0.0, dtype=tf.float32)

        for idx in range(int(args.grad_accum_steps)):
            x = xs[idx]
            y = ys[idx]
            with tf.GradientTape() as tape:
                logits = model(x, training=True)
                logits = tf.cast(logits, tf.float32)
                loss_value = tf.cast(loss_fn(y, logits), tf.float32)
                scaled_loss = loss_value / float(args.grad_accum_steps)
            grads = tape.gradient(scaled_loss, trainable)
            next_accum = []
            for current, grad, var in zip(accum_grads, grads, trainable):
                if grad is None:
                    grad = tf.zeros_like(var)
                next_accum.append(current + tf.cast(grad, current.dtype))
            accum_grads = next_accum
            total_loss = total_loss + scaled_loss

        optimizer.apply_gradients(zip(accum_grads, trainable))
        grad_norm = tf.linalg.global_norm(accum_grads)
        return total_loss, grad_norm

    compiled = tf.function(step_fn, jit_compile=bool(jit_compile)) if use_function else step_fn

    losses: list[float] = []
    grad_norms: list[float] = []
    started = time.perf_counter()
    for step in range(int(args.steps)):
        x, y = fixed_batch(tf, args, step)
        loss_value, grad_norm = compiled(x, y)
        loss_float = float(loss_value.numpy())
        grad_float = float(grad_norm.numpy())
        losses.append(loss_float)
        grad_norms.append(grad_float)
        if args.print_step_values:
            print(
                json.dumps(
                    {
                        "event": "step",
                        "arm": name,
                        "step": step,
                        "loss": loss_float,
                        "grad_norm": grad_float,
                    },
                    sort_keys=True,
                ),
                flush=True,
            )

    return {
        "arm": name,
        "use_function": bool(use_function),
        "jit_compile": bool(jit_compile),
        "final_loss": losses[-1],
        "final_grad_norm": grad_norms[-1],
        "losses": losses,
        "grad_norms": grad_norms,
        "wall_time_sec": round(time.perf_counter() - started, 4),
    }


def rel_diff(a: float, b: float) -> float:
    if not (math.isfinite(a) and math.isfinite(b)):
        return float("inf")
    return abs(a - b) / max(abs(a), abs(b), 1e-8)


def compare(name: str, ref: dict[str, Any], alt: dict[str, Any], args: argparse.Namespace) -> dict[str, Any]:
    loss_rel = rel_diff(float(ref["final_loss"]), float(alt["final_loss"]))
    grad_rel = rel_diff(float(ref["final_grad_norm"]), float(alt["final_grad_norm"]))
    divergent = bool(loss_rel > float(args.loss_rtol) or grad_rel > float(args.grad_rtol))
    return {
        "check": name,
        "status": "divergent" if divergent else "consistent",
        "loss_rel_diff": loss_rel,
        "loss_rel_threshold": float(args.loss_rtol),
        "grad_rel_diff": grad_rel,
        "grad_rel_threshold": float(args.grad_rtol),
        "ref_arm": ref["arm"],
        "alt_arm": alt["arm"],
        "ref_final_loss": ref["final_loss"],
        "alt_final_loss": alt["final_loss"],
        "ref_final_grad_norm": ref["final_grad_norm"],
        "alt_final_grad_norm": alt["final_grad_norm"],
    }


def runtime_config(tf: Any, args: argparse.Namespace) -> dict[str, Any]:
    return {
        "tensorflow_version": tf.__version__,
        "physical_gpus": [str(gpu) for gpu in tf.config.list_physical_devices("GPU")],
        "profile": args.profile,
        "steps": int(args.steps),
        "seed": int(args.seed),
        "model": "ResNet50(weights=None, include_top=False, pooling='avg') + Dense(10)",
        "input": "deterministic synthetic normal tensors",
        "image_size": int(args.image_size),
        "batch_size": int(args.batch_size),
        "grad_accum_steps": int(args.grad_accum_steps),
        "effective_batch_size": int(args.batch_size) * int(args.grad_accum_steps),
        "optimizer": "Adam",
        "learning_rate": float(args.learning_rate),
        "weight_decay": float(args.weight_decay),
        "precision_policy": str(args.precision_policy),
        "threading": {
            "intra_op_parallelism_threads": int(args.intra_op_threads),
            "inter_op_parallelism_threads": int(args.inter_op_threads),
        },
        "thresholds": {
            "loss_rel_threshold": float(args.loss_rtol),
            "grad_rel_threshold": float(args.grad_rtol),
        },
    }


def main() -> int:
    args = apply_defaults(parse_args())
    out_dir = setup_output_dir(args)
    tf = configure_runtime(args)

    checks = {part.strip() for part in str(args.checks).split(",") if part.strip()}

    # Build one deterministic template and reuse its initial weights for every arm.
    template = make_model(tf, args)
    initial_weights = template.get_weights()
    tf.keras.backend.clear_session()

    print(json.dumps({"event": "environment", **runtime_config(tf, args)}, sort_keys=True), flush=True)

    arms: dict[str, dict[str, Any]] = {}
    results: list[dict[str, Any]] = []

    if "graph_vs_eager" in checks:
        arms["graph"] = train_arm(
            tf,
            args,
            name="graph",
            use_function=True,
            jit_compile=False,
            initial_weights=initial_weights,
        )
        arms["eager"] = train_arm(
            tf,
            args,
            name="eager",
            use_function=False,
            jit_compile=False,
            initial_weights=initial_weights,
        )
        results.append(compare("graph_vs_eager", arms["graph"], arms["eager"], args))

    if "xla_vs_nonxla" in checks:
        if "nonxla" not in arms:
            arms["nonxla"] = train_arm(
                tf,
                args,
                name="nonxla",
                use_function=True,
                jit_compile=False,
                initial_weights=initial_weights,
            )
        else:
            arms["nonxla"] = arms["graph"]
        arms["xla"] = train_arm(
            tf,
            args,
            name="xla",
            use_function=True,
            jit_compile=True,
            initial_weights=initial_weights,
        )
        results.append(compare("xla_vs_nonxla", arms["nonxla"], arms["xla"], args))

    has_divergence = any(row["status"] == "divergent" for row in results)
    report = {
        "event": "summary",
        "status": "DIVERGENCE_REPRODUCED" if has_divergence else "not_reproduced",
        "has_divergence": has_divergence,
        "divergent_checks": [row["check"] for row in results if row["status"] == "divergent"],
        "runtime_config": runtime_config(tf, args),
        "arms": arms,
        "results": results,
    }

    (out_dir / "report.json").write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    (out_dir / "summary.json").write_text(json.dumps(report, sort_keys=True) + "\n", encoding="utf-8")
    print(json.dumps(report, sort_keys=True), flush=True)
    print(f"output_dir={out_dir}", flush=True)

    if args.fail_on_divergence and has_divergence:
        return 2
    return 0

if __name__ == "__main__":
    raise SystemExit(main())
```

### Relevant log output

```shell
| arm | step 1 loss | step 20 loss | step 1 grad norm | step 20 grad norm |
|---|---:|---:|---:|---:|
| graph | 4.083001136779785 | 5.515571594238281 | 853.9214477539062 | 8.42819881439209 |
| eager | 4.082990646362305 | 3.878284454345703 | 853.6309204101562 | 4.26118278503418 |
```

**Comment by CodersAcademy006** [comment]
Hey @whitephilomel, Thanks for the repro again; this one's worth a closer look.

Step 1 is basically in agreement (4.083001 vs 4.082990, ~1e-5 relative diff), but the grad norm at step 1 is ~854 for both arms, which puts this training run in a near-unstable regime (`lr=0.05` Adam on ResNet50 from scratch, no warmup). That kind of regime has a positive Lyapunov exponent, so tiny perturbations get amplified exponentially rather than staying bounded.

My guess: even with `jit_compile=False`, `tf.function` still applies Grappler graph optimizations (arithmetic optimizer, remapper, layout optimizer, CSE) that don't run in eager. These can reorder float32 ops, producing a ULP-scale difference at step 1, which then compounds over 20 steps through the unstable loss landscape into a meaningfully different trajectory by step 20. Neither run is invalid (no NaN/Inf), they've just diverged into different points on a chaotic surface.

Two quick checks that would confirm this:

- Disable Grappler (`tf.config.optimizer.set_experimental_options({'arithmetic_optimization': False, 'remapping': False, 'layout_optimizer': False, 'constant_folding': False})`) and re-run. If step 1 shrinks to true bitwise/ULP and the pattern holds, that's the seed.
- Re-run at a standard ResNet50 LR (e.g. `1e-3` with warmup, grad norms in single digits). If the 20-step divergence collapses, that confirms chaotic amplification rather than a correctness bug.
